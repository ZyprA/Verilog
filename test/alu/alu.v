`define ADD 5'b00000
`define SUB 5'b00001
`define MUL 5'b00010
`define SHL 5'b00011
`define SHR 5'b00100
`define BAND 5'b00101
`define BOR 5'b00110
`define BXOR 5'b00111
`define AND 5'b01000
`define OR 5'b01001
`define EQ 5'b01010
`define NE 5'b01011
`define GE 5'b01100
`define LE 5'b01101
`define GT 5'b01110
`define LT 5'b01111
`define NEG 5'b10000
`define BNOT 5'b10001
`define NOT 5'b10010

module ArithmeticLogicUnit(a, b, f, s);
    input [15:0] a, b;
    input [4:0] f;
    output [15:0] s;

    wire [15:0] x, y;

    assign x = a + 16'h8000;
    assign y = b + 16'h8000;

    function [15:0] alu;
        input [15:0] a, b;
        input [4:0] f;
        input [15:0] x, y;
        
        case (f)
            `ADD : alu = b + a;
            `SUB : alu = b - a;
            `MUL : alu = b * a;
            `SHL : alu = b << a;
            `SHR : alu = b >> a;
            `BAND : alu = b & a;
            `BOR : alu = a | b;
            `AND : alu = b && a;
            `OR : alu = b || a;
            `EQ : alu = b == a;
            `NE : alu = b != a;
            `GE : alu = x >= y;
            `LE : alu = x <= y;
            `GT : alu = x > y;
            `LT : alu = x < y;
            `NEG : alu = -a;
            `BNOT : alu = ~a;
            `NOT : alu = !a;
            default : alu = 16'hxxxx;
        endcase
    endfunction

    assign s = alu(a, b, f, x, y);

endmodule

module alu_sim;
    reg [15:0] a, b;
    reg [4:0] f;
    wire [15:0] s;

    ArithmeticLogicUnit alu(a,b,f,s);

    initial begin
        $dumpfile("alu.vcd");
        $dumpvars(0, alu_sim);
    end

    initial begin
        a = 10; b = 40;
        #100 f = `ADD;
        #100 f = `SUB;
        #100 f = `MUL;
        #100 f = `SHL;
        #100 f = `SHR;
        #100 f = `BAND;
        #100 f = `BOR;
        #100 f = `AND;
        #100 f = `OR;
        #100 f = `EQ;
        #100 f = `NE;
        #100 f = `GE;
        #100 f = `LE;
        #100 f = `GT;
        #100 f = `LT;
        #100 f = `NEG;
        #100 f = `BNOT;
        #100 f = `NOT;
        #100 $finish;

    end

endmodule