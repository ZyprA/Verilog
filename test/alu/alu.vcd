$date
	Sun Feb  2 16:08:16 2025
$end
$version
	Icarus Verilog
$end
$timescale
	1s
$end
$scope module alu_sim $end
$var wire 16 ! s [15:0] $end
$var reg 16 " a [15:0] $end
$var reg 16 # b [15:0] $end
$var reg 5 $ f [4:0] $end
$scope module alu $end
$var wire 16 % a [15:0] $end
$var wire 16 & b [15:0] $end
$var wire 5 ' f [4:0] $end
$var wire 16 ( y [15:0] $end
$var wire 16 ) x [15:0] $end
$var wire 16 * s [15:0] $end
$scope function alu $end
$var reg 16 + a [15:0] $end
$var reg 16 , b [15:0] $end
$var reg 5 - f [4:0] $end
$var reg 16 . x [15:0] $end
$var reg 16 / y [15:0] $end
$upscope $end
$upscope $end
$upscope $end
$enddefinitions $end
$comment Show the parameter values. $end
$dumpall
$end
#0
$dumpvars
b1000000000101000 /
b1000000000001010 .
bx -
b101000 ,
b1010 +
bx *
b1000000000001010 )
b1000000000101000 (
bx '
b101000 &
b1010 %
bx $
b101000 #
b1010 "
bx !
$end
#100
b110010 !
b110010 *
b0 -
b0 $
b0 '
#200
b11110 !
b11110 *
b1 -
b1 $
b1 '
#300
b110010000 !
b110010000 *
b10 -
b10 $
b10 '
#400
b1010000000000000 !
b1010000000000000 *
b11 -
b11 $
b11 '
#500
b0 !
b0 *
b100 -
b100 $
b100 '
#600
b1000 !
b1000 *
b101 -
b101 $
b101 '
#700
b101010 !
b101010 *
b110 -
b110 $
b110 '
#800
b1 !
b1 *
b1000 -
b1000 $
b1000 '
#900
b1001 -
b1001 $
b1001 '
#1000
b0 !
b0 *
b1010 -
b1010 $
b1010 '
#1100
b1 !
b1 *
b1011 -
b1011 $
b1011 '
#1200
b0 !
b0 *
b1100 -
b1100 $
b1100 '
#1300
b1 !
b1 *
b1101 -
b1101 $
b1101 '
#1400
b0 !
b0 *
b1110 -
b1110 $
b1110 '
#1500
b1 !
b1 *
b1111 -
b1111 $
b1111 '
#1600
b1111111111110110 !
b1111111111110110 *
b10000 -
b10000 $
b10000 '
#1700
b1111111111110101 !
b1111111111110101 *
b10001 -
b10001 $
b10001 '
#1800
b0 !
b0 *
b10010 -
b10010 $
b10010 '
#1900
