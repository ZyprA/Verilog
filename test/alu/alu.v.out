#! /opt/homebrew/Cellar/icarus-verilog/12.0/bin/vvp
:ivl_version "12.0 (stable)";
:ivl_delay_selection "TYPICAL";
:vpi_time_precision + 0;
:vpi_module "/opt/homebrew/Cellar/icarus-verilog/12.0/lib/ivl/system.vpi";
:vpi_module "/opt/homebrew/Cellar/icarus-verilog/12.0/lib/ivl/vhdl_sys.vpi";
:vpi_module "/opt/homebrew/Cellar/icarus-verilog/12.0/lib/ivl/vhdl_textio.vpi";
:vpi_module "/opt/homebrew/Cellar/icarus-verilog/12.0/lib/ivl/v2005_math.vpi";
:vpi_module "/opt/homebrew/Cellar/icarus-verilog/12.0/lib/ivl/va_math.vpi";
S_0x154704170 .scope module, "alu_sim" "alu_sim" 2 63;
 .timescale 0 0;
v0x154715170_0 .var "a", 15 0;
v0x154715220_0 .var "b", 15 0;
v0x1547152d0_0 .var "f", 4 0;
v0x1547153a0_0 .net "s", 15 0, L_0x154715730;  1 drivers
S_0x1547042f0 .scope module, "alu" "ArithmeticLogicUnit" 2 68, 2 21 0, S_0x154704170;
 .timescale 0 0;
    .port_info 0 /INPUT 16 "a";
    .port_info 1 /INPUT 16 "b";
    .port_info 2 /INPUT 5 "f";
    .port_info 3 /OUTPUT 16 "s";
L_0x148040010 .functor BUFT 1, C4<1000000000000000>, C4<0>, C4<0>, C4<0>;
v0x154714b70_0 .net/2u *"_ivl_0", 15 0, L_0x148040010;  1 drivers
L_0x148040058 .functor BUFT 1, C4<1000000000000000>, C4<0>, C4<0>, C4<0>;
v0x154714c30_0 .net/2u *"_ivl_4", 15 0, L_0x148040058;  1 drivers
v0x154714cd0_0 .net "a", 15 0, v0x154715170_0;  1 drivers
v0x154714d80_0 .net "b", 15 0, v0x154715220_0;  1 drivers
v0x154714e30_0 .net "f", 4 0, v0x1547152d0_0;  1 drivers
v0x154714f20_0 .net "s", 15 0, L_0x154715730;  alias, 1 drivers
v0x154714fd0_0 .net "x", 15 0, L_0x154715490;  1 drivers
v0x154715080_0 .net "y", 15 0, L_0x1547155f0;  1 drivers
L_0x154715490 .arith/sum 16, v0x154715170_0, L_0x148040010;
L_0x1547155f0 .arith/sum 16, v0x154715220_0, L_0x148040058;
L_0x154715730 .ufunc/vec4 TD_alu_sim.alu.alu, 16, v0x154715170_0, v0x154715220_0, v0x1547152d0_0, L_0x154715490, L_0x1547155f0 (v0x1547046f0_0, v0x154714860_0, v0x154714920_0, v0x1547149d0_0, v0x154714ac0_0) S_0x154704520;
S_0x154704520 .scope function.vec4.s16, "alu" "alu" 2 31, 2 31 0, S_0x1547042f0;
 .timescale 0 0;
v0x1547046f0_0 .var "a", 15 0;
; Variable alu is vec4 return value of scope S_0x154704520
v0x154714860_0 .var "b", 15 0;
v0x154714920_0 .var "f", 4 0;
v0x1547149d0_0 .var "x", 15 0;
v0x154714ac0_0 .var "y", 15 0;
TD_alu_sim.alu.alu ;
    %load/vec4 v0x154714920_0;
    %dup/vec4;
    %pushi/vec4 0, 0, 5;
    %cmp/u;
    %jmp/1 T_0.0, 6;
    %dup/vec4;
    %pushi/vec4 1, 0, 5;
    %cmp/u;
    %jmp/1 T_0.1, 6;
    %dup/vec4;
    %pushi/vec4 2, 0, 5;
    %cmp/u;
    %jmp/1 T_0.2, 6;
    %dup/vec4;
    %pushi/vec4 3, 0, 5;
    %cmp/u;
    %jmp/1 T_0.3, 6;
    %dup/vec4;
    %pushi/vec4 4, 0, 5;
    %cmp/u;
    %jmp/1 T_0.4, 6;
    %dup/vec4;
    %pushi/vec4 5, 0, 5;
    %cmp/u;
    %jmp/1 T_0.5, 6;
    %dup/vec4;
    %pushi/vec4 6, 0, 5;
    %cmp/u;
    %jmp/1 T_0.6, 6;
    %dup/vec4;
    %pushi/vec4 8, 0, 5;
    %cmp/u;
    %jmp/1 T_0.7, 6;
    %dup/vec4;
    %pushi/vec4 9, 0, 5;
    %cmp/u;
    %jmp/1 T_0.8, 6;
    %dup/vec4;
    %pushi/vec4 10, 0, 5;
    %cmp/u;
    %jmp/1 T_0.9, 6;
    %dup/vec4;
    %pushi/vec4 11, 0, 5;
    %cmp/u;
    %jmp/1 T_0.10, 6;
    %dup/vec4;
    %pushi/vec4 12, 0, 5;
    %cmp/u;
    %jmp/1 T_0.11, 6;
    %dup/vec4;
    %pushi/vec4 13, 0, 5;
    %cmp/u;
    %jmp/1 T_0.12, 6;
    %dup/vec4;
    %pushi/vec4 14, 0, 5;
    %cmp/u;
    %jmp/1 T_0.13, 6;
    %dup/vec4;
    %pushi/vec4 15, 0, 5;
    %cmp/u;
    %jmp/1 T_0.14, 6;
    %dup/vec4;
    %pushi/vec4 16, 0, 5;
    %cmp/u;
    %jmp/1 T_0.15, 6;
    %dup/vec4;
    %pushi/vec4 17, 0, 5;
    %cmp/u;
    %jmp/1 T_0.16, 6;
    %dup/vec4;
    %pushi/vec4 18, 0, 5;
    %cmp/u;
    %jmp/1 T_0.17, 6;
    %pushi/vec4 65535, 65535, 16;
    %ret/vec4 0, 0, 16;  Assign to alu (store_vec4_to_lval)
    %jmp T_0.19;
T_0.0 ;
    %load/vec4 v0x154714860_0;
    %load/vec4 v0x1547046f0_0;
    %add;
    %ret/vec4 0, 0, 16;  Assign to alu (store_vec4_to_lval)
    %jmp T_0.19;
T_0.1 ;
    %load/vec4 v0x154714860_0;
    %load/vec4 v0x1547046f0_0;
    %sub;
    %ret/vec4 0, 0, 16;  Assign to alu (store_vec4_to_lval)
    %jmp T_0.19;
T_0.2 ;
    %load/vec4 v0x154714860_0;
    %load/vec4 v0x1547046f0_0;
    %mul;
    %ret/vec4 0, 0, 16;  Assign to alu (store_vec4_to_lval)
    %jmp T_0.19;
T_0.3 ;
    %load/vec4 v0x154714860_0;
    %ix/getv 4, v0x1547046f0_0;
    %shiftl 4;
    %ret/vec4 0, 0, 16;  Assign to alu (store_vec4_to_lval)
    %jmp T_0.19;
T_0.4 ;
    %load/vec4 v0x154714860_0;
    %ix/getv 4, v0x1547046f0_0;
    %shiftr 4;
    %ret/vec4 0, 0, 16;  Assign to alu (store_vec4_to_lval)
    %jmp T_0.19;
T_0.5 ;
    %load/vec4 v0x154714860_0;
    %load/vec4 v0x1547046f0_0;
    %and;
    %ret/vec4 0, 0, 16;  Assign to alu (store_vec4_to_lval)
    %jmp T_0.19;
T_0.6 ;
    %load/vec4 v0x1547046f0_0;
    %load/vec4 v0x154714860_0;
    %or;
    %ret/vec4 0, 0, 16;  Assign to alu (store_vec4_to_lval)
    %jmp T_0.19;
T_0.7 ;
    %load/vec4 v0x154714860_0;
    %cmpi/ne 0, 0, 16;
    %flag_get/vec4 4;
    %jmp/0 T_0.20, 4;
    %load/vec4 v0x1547046f0_0;
    %pushi/vec4 0, 0, 16;
    %cmp/ne;
    %flag_get/vec4 4;
    %and;
T_0.20;
    %pad/u 16;
    %ret/vec4 0, 0, 16;  Assign to alu (store_vec4_to_lval)
    %jmp T_0.19;
T_0.8 ;
    %load/vec4 v0x154714860_0;
    %cmpi/ne 0, 0, 16;
    %flag_get/vec4 4;
    %jmp/1 T_0.21, 4;
    %load/vec4 v0x1547046f0_0;
    %pushi/vec4 0, 0, 16;
    %cmp/ne;
    %flag_get/vec4 4;
    %or;
T_0.21;
    %pad/u 16;
    %ret/vec4 0, 0, 16;  Assign to alu (store_vec4_to_lval)
    %jmp T_0.19;
T_0.9 ;
    %load/vec4 v0x154714860_0;
    %load/vec4 v0x1547046f0_0;
    %cmp/e;
    %flag_get/vec4 4;
    %pad/u 16;
    %ret/vec4 0, 0, 16;  Assign to alu (store_vec4_to_lval)
    %jmp T_0.19;
T_0.10 ;
    %load/vec4 v0x154714860_0;
    %load/vec4 v0x1547046f0_0;
    %cmp/ne;
    %flag_get/vec4 4;
    %pad/u 16;
    %ret/vec4 0, 0, 16;  Assign to alu (store_vec4_to_lval)
    %jmp T_0.19;
T_0.11 ;
    %load/vec4 v0x154714ac0_0;
    %load/vec4 v0x1547149d0_0;
    %cmp/u;
    %flag_get/vec4 4;
    %flag_get/vec4 5;
    %or;
    %pad/u 16;
    %ret/vec4 0, 0, 16;  Assign to alu (store_vec4_to_lval)
    %jmp T_0.19;
T_0.12 ;
    %load/vec4 v0x1547149d0_0;
    %load/vec4 v0x154714ac0_0;
    %cmp/u;
    %flag_get/vec4 4;
    %flag_get/vec4 5;
    %or;
    %pad/u 16;
    %ret/vec4 0, 0, 16;  Assign to alu (store_vec4_to_lval)
    %jmp T_0.19;
T_0.13 ;
    %load/vec4 v0x154714ac0_0;
    %load/vec4 v0x1547149d0_0;
    %cmp/u;
    %flag_get/vec4 5;
    %pad/u 16;
    %ret/vec4 0, 0, 16;  Assign to alu (store_vec4_to_lval)
    %jmp T_0.19;
T_0.14 ;
    %load/vec4 v0x1547149d0_0;
    %load/vec4 v0x154714ac0_0;
    %cmp/u;
    %flag_get/vec4 5;
    %pad/u 16;
    %ret/vec4 0, 0, 16;  Assign to alu (store_vec4_to_lval)
    %jmp T_0.19;
T_0.15 ;
    %load/vec4 v0x1547046f0_0;
    %inv;
    %pushi/vec4 1, 0, 16;
    %add;
    %ret/vec4 0, 0, 16;  Assign to alu (store_vec4_to_lval)
    %jmp T_0.19;
T_0.16 ;
    %load/vec4 v0x1547046f0_0;
    %inv;
    %ret/vec4 0, 0, 16;  Assign to alu (store_vec4_to_lval)
    %jmp T_0.19;
T_0.17 ;
    %load/vec4 v0x1547046f0_0;
    %nor/r;
    %pad/u 16;
    %ret/vec4 0, 0, 16;  Assign to alu (store_vec4_to_lval)
    %jmp T_0.19;
T_0.19 ;
    %pop/vec4 1;
    %end;
    .scope S_0x154704170;
T_1 ;
    %vpi_call 2 71 "$dumpfile", "alu.vcd" {0 0 0};
    %vpi_call 2 72 "$dumpvars", 32'sb00000000000000000000000000000000, S_0x154704170 {0 0 0};
    %end;
    .thread T_1;
    .scope S_0x154704170;
T_2 ;
    %pushi/vec4 10, 0, 16;
    %store/vec4 v0x154715170_0, 0, 16;
    %pushi/vec4 40, 0, 16;
    %store/vec4 v0x154715220_0, 0, 16;
    %delay 100, 0;
    %pushi/vec4 0, 0, 5;
    %store/vec4 v0x1547152d0_0, 0, 5;
    %delay 100, 0;
    %pushi/vec4 1, 0, 5;
    %store/vec4 v0x1547152d0_0, 0, 5;
    %delay 100, 0;
    %pushi/vec4 2, 0, 5;
    %store/vec4 v0x1547152d0_0, 0, 5;
    %delay 100, 0;
    %pushi/vec4 3, 0, 5;
    %store/vec4 v0x1547152d0_0, 0, 5;
    %delay 100, 0;
    %pushi/vec4 4, 0, 5;
    %store/vec4 v0x1547152d0_0, 0, 5;
    %delay 100, 0;
    %pushi/vec4 5, 0, 5;
    %store/vec4 v0x1547152d0_0, 0, 5;
    %delay 100, 0;
    %pushi/vec4 6, 0, 5;
    %store/vec4 v0x1547152d0_0, 0, 5;
    %delay 100, 0;
    %pushi/vec4 8, 0, 5;
    %store/vec4 v0x1547152d0_0, 0, 5;
    %delay 100, 0;
    %pushi/vec4 9, 0, 5;
    %store/vec4 v0x1547152d0_0, 0, 5;
    %delay 100, 0;
    %pushi/vec4 10, 0, 5;
    %store/vec4 v0x1547152d0_0, 0, 5;
    %delay 100, 0;
    %pushi/vec4 11, 0, 5;
    %store/vec4 v0x1547152d0_0, 0, 5;
    %delay 100, 0;
    %pushi/vec4 12, 0, 5;
    %store/vec4 v0x1547152d0_0, 0, 5;
    %delay 100, 0;
    %pushi/vec4 13, 0, 5;
    %store/vec4 v0x1547152d0_0, 0, 5;
    %delay 100, 0;
    %pushi/vec4 14, 0, 5;
    %store/vec4 v0x1547152d0_0, 0, 5;
    %delay 100, 0;
    %pushi/vec4 15, 0, 5;
    %store/vec4 v0x1547152d0_0, 0, 5;
    %delay 100, 0;
    %pushi/vec4 16, 0, 5;
    %store/vec4 v0x1547152d0_0, 0, 5;
    %delay 100, 0;
    %pushi/vec4 17, 0, 5;
    %store/vec4 v0x1547152d0_0, 0, 5;
    %delay 100, 0;
    %pushi/vec4 18, 0, 5;
    %store/vec4 v0x1547152d0_0, 0, 5;
    %delay 100, 0;
    %vpi_call 2 95 "$finish" {0 0 0};
    %end;
    .thread T_2;
# The file index is used to find the file name in the following table.
:file_names 3;
    "N/A";
    "<interactive>";
    "alu.v";
